[sym_shape][perf] eval_static: guards, unbacked compute once

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124217

Differential Revision: [D56212345](https://our.internmc.facebook.com/intern/diff/D56212345)

IvanKobzarev committed:
[sym_shape][perf] eval_static: guards, unbacked compute once

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3623,8 +3641,8 @@ def add_expr(expr):
 
-            # NB: this helps us deal with And/Or connectives
-            expr = expr.subs(subst)
+		# NB: this helps us deal with And/Or connectives
```
tabs

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3593,6 +3593,27 @@ def _maybe_evaluate_static(
         # Apply known runtime asserts
+        guards_exprs = []
+        for g in self.guards:
+            e = self.simplify(g.expr)
```
On master, we're not calling simplify yet. I guess it's harmless enough to do here though.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3593,6 +3593,27 @@ def _maybe_evaluate_static(
+                e = canonicalize_bool_expr(e.xreplace(self.var_to_val))
+            guards_exprs.append(e)
+
+        symbols_unbacked = symbols - self.var_to_val.keys()
```
Isn't this set on symbols, aren't these slow?

IvanKobzarev replied:
Yes, this is similar to that strange optimization, but profile did not show this line

ezyang replied:
yeah, this is partially why I feel there was something very strange going on with that particular optimization.

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56212345).

IvanKobzarev committed:
Update on "[sym_shape][perf] eval_static: guards, unbacked compute once"


Differential Revision: [D56212345](https://our.internmc.facebook.com/intern/diff/D56212345)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56212345).

lezcano commented:
Another approach to the same issue: https://github.com/pytorch/pytorch/pull/123342/

IvanKobzarev committed:
Update on "[sym_shape][perf] eval_static: guards, unbacked compute once"


Differential Revision: [D56212345](https://our.internmc.facebook.com/intern/diff/D56212345)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[sym_shape][perf] eval_static: guards, unbacked compute once"


Differential Revision: [D56212345](https://our.internmc.facebook.com/intern/diff/D56212345)

[ghstack-poisoned]

IvanKobzarev commented:
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56212345).
